### DISCUSSION TITLE
大佬可以出一期WPF单页应用的视频吗？

### DISCUSSION THREAD

**Discussion by CrazyZhang666** [BYJRK/DotNet-Discussions#4] [Q&A]
类似下面这样，左右布局，左边是各种页面的切换按钮，右边是页面展示区（用来展示切换的页面）

![image](https://user-images.githubusercontent.com/28080853/211270886-1e0c6b14-82a1-4f85-b033-8b279b77f4c0.png)

我目前只会这样切换（使用WPF原生代码，不使用第三方框架，比如Prism）

# 左边
左边使用RaddionButtom按钮组或者ListBox（自定义样式），使用Command命令绑定，使用CommandParameter来传递页面切换参数

![image](https://user-images.githubusercontent.com/28080853/211271771-8fec0ec4-6186-49c1-bba5-93ee12e195a7.png)

# 右边

右边是使用ContentControl来切换页面

![image](https://user-images.githubusercontent.com/28080853/211271870-bb9fb889-edeb-4854-9180-59093d99c6e6.png)

页面使用用户控件，先实例化好

![image](https://user-images.githubusercontent.com/28080853/211272176-76b71cd2-2d35-45e6-9984-8c40fe298b14.png)

然后使用CommandParameter传递的参数切换对应页面

![image](https://user-images.githubusercontent.com/28080853/211272147-e9c36bc5-b899-47b9-bbe7-865a831d8f94.png)

# 额外请求

Winform也可以实现类似的效果吗？要具体怎么实现呢？

**Comment by BYJRK** [comment]
看起来不错，其实实现方法有很多，比如TabControl，Frame，还有UWP新增的NavigationView等。对应的View还需要搭配ViewModel，可以直接放在MainWindowViewModel里面实例化，也可以交给IoC等。WinForms我不是很清楚，主要是我没用过WinForms开发复杂点的软件，以前开发过的都是非常简单的小工具什么的

**Comment by Hakoyu** [comment]
我推荐是使用ListBox来达到互斥效果,并使用Frame显示
```csharp
Dictionary<string, Page> pages = new();

Page1 page1 = new();
Page2 page2 = new();

pages.Add(nameof(Page1), page1);
pages.Add(nameof(Page2), page2);

ListBox.Items.Add(CreateListBoxItemForPage(nameof(Page1), nameof(Page1), page1));
ListBox.Items.Add(CreateListBoxItemForPage(nameof(Page2), nameof(Page2), page2));

ListBoxItem CreateListBoxItemForPage(string name, string id)
    => new ListBoxItem()
    {
        Content = name,
        Tag = id,
    };

void ListBox_SelectionChanged(object sender, SelectionChangedEventArgs e)
{
    if (sender is ListBox listBox && listBox.SelectedIndex != -1 && listBox.SelectedItem is ListBoxItem item)
    {
        var id = item.Tag.ToString()!;
        Frame.Content = pages[id];
    }
}
```
当然 实际工程中不推荐直接实例化 而是使用Lazy来进行延迟加载,以提高性能
![Honeycam 2023-01-11 18-31-04](https://user-images.githubusercontent.com/48956359/211783599-7a766c13-d341-4b52-ad63-27bf8f3d3917.gif)

**Reply by BYJRK** [reply]
你的这个点子也不错。如果是MVVM呢？其实如果只是互斥效果的话，而且还用Frame，直接用NavigationView吧，比如[ModernWpf](https://github.com/Kinnara/ModernWpf)提供的

**Reply by Hakoyu** [reply]
是的 方法有很多种
一个一个判断的话实在有点傻

**Comment by KiyanYang** [comment]
可以尝试一下 https://github.com/microsoft/TemplateStudio 这个项目模板，其中的 Navigation Pane 项目类型应该符合你的要求。
